### PR TITLE
documentation: index: Add the documentation for assignment 2

### PR DESCRIPTION
**Pull request by consra**
Add SO2 assignment 2 to git repo. Original is at:
https://ocw.cs.pub.ro/courses/so2/teme/tema2

This assignment requests from the students to implement a
driver for the serial port (UART16550).

Signed-off-by: Constantin Raducanu <raducanu.costi@gmail.com>

### CODE REVIEW

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+-  Deadline: :command:`Wednesday, 21 April 2021, 23:00`
+
+Assignment's Objectives
+=======================
+
+*  consolidating the knowledge of device driver
+*  read hardware documentation and track the desired functionality in the documentation
+*  work with interruptions; use of non-blocking functions in interrupt context
+*  use of buffers; synchronization
```
Comment (dbaluta) [inline Documentation/teaching/so2/assign2-driver-uart.rst:13]: work with interrupts.

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 2]
```diff
@@ -0,0 +1,163 @@
+==========================
+Assignment 2 - Driver UART
+==========================
+
+-  Deadline: :command:`Wednesday, 21 April 2021, 23:00`
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:5]: April 21st

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+Assignment's Objectives
+=======================
+
+*  consolidating the knowledge of device driver
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:10]: drivers - plural

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+*  read hardware documentation and track the desired functionality in the documentation
+*  work with interrupts; use of non-blocking functions in interrupt context
+*  use of buffers; synchronization
+*  kernel modules with parameters
+
+Statement
+=========
+
+Write a kernel module that implements a driver for the serial port (`UART16550`).
+The device driver must support the two standard serial ports in a PC, `COM1` and `COM2` (`0x3f8` and `0x2f8`,
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:20]: close the parenthesis

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+in fact the entire range of `8` addresses `0x3f8-0x3ff` and `0x2f8-0x2ff` specific to the two ports).
+In addition to the standard routines (`open`, `read`, `write`, `close`),
+the driver must also have support for changing communication parameters using an `ioctl` operation (`UART16550_IOCTL_SET_LINE`).
+
+The driver must use interrupts for both reception and transmission to reduce latency and CPU usage time.
+`Read` and `write` calls must also be blocking. :command:`Assignments that do not meet these requirements will not be considered.`
+It is recommended that you use a buffer for read and another buffer for write buffer for each serial port in the driver.
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:27]: rephrase: ...and another buffer for write buffer...

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+(the write buffer in the kernel is full and no data can be written).
+
+Buffers Scheme
+--------------
+
+.. image:: ../img/buffers-scheme.png
+
+Data transfer between the various buffers is a `Producer-Consumer <https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem>`__ problem. Example:
+
+-   if it is written from a process to the device, the process is the producer and the device is the consumer; the process will block until there is at least one free space in the consumer's buffer
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:41]: rephrase: - if a process writes to the device the process is the producer...
same for the next sentence, with the read case

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+   -  the `major` parameter will specify the major with which the device must be registered
+   -  the `option` parameter will specify how it works:
+
+      -  OPTION_BOTH: will also register COM1 and COM2, with the major given by the `major` parameter and the minors 0 (for COM1) and 1 (for COM2);
+      -  OPTION_COM1: will only register COM1, with the major `major` and minor 0;
+      -  OPTION_COM2: will only register COM2, with the major `major` and minor 1;
+   -  to learn how to pass parameters in Linux, see `tldp <https://tldp.org/LDP/lkmpg/2.6/html/x323.html>`__
+   -  the default values are `major=42` and `option=OPTION_BOTH`.
+-  the number of interrupts associated with COM1 is 4 (`IRQ_COM1`) and the number of interrupts associated with COM2 is 3 (`IRQ_COM2`)
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:59]: Here it's actually the "interrupt number".
So it becomes: - the interrupt number associated with COM1 is 4 ...

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+-  `the header <elf.cs.pub.ro/so2/res/teme/uart16550.h>`__ with the definitions needed for special operations;
+-  a starting point in implementing read / write routines is the `example <elf.cs.pub.ro/so2/res/teme/case.zip>`__ of uppercase / lowercase character device driver; the only difference is that you have to use two buffers, one for read and one for write;
+-  you can use `kfifo <elf.cs.pub.ro/so2/res/teme/case.zip>`__ for buffers;
+-  you do not have to use deferred functions to read / write data from / to ports (you can do everything from interruption context);
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:63]: Replace interruption context with "interrupt context"

File: Documentation/teaching/so2/assign2-driver-uart.rst [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+-  you will need to synchronize the read / write routines with the interrupt handling routine for the routines to be blocking; it is recommended to use `synchronization with queues <https://ocw.cs.pub.ro/courses/so2/laboratoare/lab04?&#sincronizare_-_cozi_de_asteptare>`__
```
Comment (claudiughioc) [inline Documentation/teaching/so2/assign2-driver-uart.rst:64]: Use "synchronization with waiting queues"

File: Documentation/teaching/so2/assign2-driver-uart.rst (hunk 2 above)
Comment (claudiughioc) [reply Documentation/teaching/so2/assign2-driver-uart.rst:5]: Also, add a note mentioning that this assignment can be made in teams (max 2). Only one of them must submit the assignment, and the names of the student should be listed in a README file.

**Review by danield20 — APPROVED** [review]

**Review by claudiughioc — APPROVED** [review]

**Pull request merged**